### ISSUE TITLE
RecursionError: maximum recursion depth exceeded in self.search loading training data

### ISSUE THREAD

**Issue by L-ENA** [dedupeio/dedupe#876]
Hello!

I'm using dedupe 2.0.6  on Windows with python 3.6. (and dedupe_variable_name + datetime packages, although not using in examlple below), I'm lucky enough to have a rather large set of labelled data for my project, but whenever I try to train with too many records then I'm getting this recursion error below. The stacktrace was produced using a training JSON file with 952 records (equalling 476 matched pairs) and the code from csv_example.py

The beginning of my JSOn training file is below the stack trace , as example for the data.

The error appears when dedupe tries to read the labelled examples. If I limit the number of examples, then it works.  I tried reducing the number of fields, and changing field types. 

Related problem: With a reduced amount of training data I am getting very good results evaluating on random samples of my dataset with around 2000 rows. But if I try and evaluate on 7-10,000 rows then the number of duplicates detected by dedupe reduces quite dramatically. Has this happened to anyone else, or is it possible that I am maing a mistake? I'm using the csv example and evaluation code.
Many thanks!

This is the recursion error:
`

Traceback (most recent call last):
File "C:/Users/lena.schmidt/PycharmProjects/dedupe/dedupe-examples/csv_example/csv_example.py", line 130, in <module> deduper.prepare_training(data_d, f)

File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\api.py", line 1249, in prepare_training
    self._sample(data, sample_size, blocked_proportion, original_length)

File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\api.py", line 1289, in _sample
    self.active_learner.mark(examples, y)

File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\labeler.py", line 362, in mark
    learner.fit_transform(self.pairs, self.y)
  File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\labeler.py", line 198, in fit_transform
    recall=1.0)

File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\training.py", line 48, in learn
    final_predicates = searcher.search(dupe_cover)

File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\training.py", line 275, in search
    self.search(reduced, partial)

File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\training.py", line 275, in search
    self.search(reduced, partial)

File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\training.py", line 275, in search
    self.search(reduced, partial)
  [Previous line repeated 982 more times]

File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\training.py", line 271, in search
    self.search(remaining, partial + (best,))

File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\training.py", line 246, in search
    covered = self.covered(partial)

File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\training.py", line 291, in covered
    for p in partial)))

File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\training.py", line 291, in <genexpr>
    for p in partial)))

  File "C:\Users\lena.schmidt\AppData\Local\conda\conda\envs\deduping\lib\site-packages\dedupe\training.py", line 398, in __getitem__
    return self._d[k]
RecursionError: maximum recursion depth exceeded`

The JSON, beginning:

{"distinct": [], "match": [[{"title": "role of erythropoietin in 7,12-dimethylbenz(a)anthracene induction of acute chromosome aberration and leukemia in rat", "url": "<go to isi>://wos:a1971k842500034", "year": "1971.0", "issn": "00278424", "volume": "68", "abstract": null, "doi": "10.1073/pnas.68.11.2761", "number": "11", "journal": "proceedings of the national academy of sciences of the united states of america", "authors": "sugiyama, t", "source": "other", "True Id": 6, "Id": 14}, {"title": "role of erythropoietin in 7,12-dimethylbenz(a)anthracene induction of acute chromosome aberration and leukemia in the rat", "url": null, "year": "1971.0", "issn": "00278424", "volume": "68", "abstract": "the incidence of chromosome aberrations  [... abstract omitted due to copyright]", "doi": "10.1073/pnas.68.11.2761", "number": "11", "journal": "proceedings of the national academy of sciences of the united states of america", "authors": "sugiyama, t", "source": "nlm", "True Id": 6, "Id": 15}],......

**Comment by fgregg** [comment]
hmm.. that's very strange. 

can you load you json file directly like

```python
import json
with open('path to your file') as f:
    data = json.load(file)
```

**Comment by L-ENA** [comment]
Hi!
Thanks for the quick reply, did not see it until now. I managed to train a good model, so this is not urgent, but using only a fraction of my training data. if I try to use more training data then the recursion error persists.

@fgregg Re: your comment, thanks for the suggestion. Yes I can load it like that, and even print the contents, they look fine. 

I've uploaded a version of the json file that does throw the error (with some 2000 records in around 1000 matched training pairs) here: https://github.com/L-ENA/misc_data
It loads rather long until it throws the error.

I've also added a smaller file (created with the same script but only 586 records/half as many dupe pairs) and this file works fine. With decreasing sizes of training files I get better evaluation results, with about 450 records/225 dupe pairs being ideal to train the best performing model...

Anyway, if this is an oversight/mistake on my side then I am really sorry..
Lena
